text/0000-ghost-busting.md

```diff
@@ -0,0 +1,732 @@
+```
+
+or equivalently:
+
+```rust
+struct Label<T>(PhantomData<T>);
+```
+
+This is a pain point since users must now:
+
+1. import `std::marker::PhantomData`,
+
+2. add a field of type [`PhantomData<T>`],
+
+3. acknowledge the field by adding an expression `PhantomData`
+   when a value of the type is to be constructed.
+
+These steps add unnecessary boilerplate and is in the way of rapid prototyping
+and developer flow when you are removing and adding fields.
+
+By making `struct Label<T>;` be owning and covariant by default in `T`,
+ergonomics of generics and label types are improved and `PhantomData<T>`
+becomes just a library type.
+
+For types where you really need to change the variance and drop checking
+behavior such as for the following `MyVec<T>` type, we gain in ergonomics
+by being able to skip steps 1 and 3 and instead write:
+
+```rust
+struct MyVec<T> {
+    phantom T,
+    data: *const T,
+    len: usize,
+    cap: usize,
+}
+```
+
+## More permissive automatic deriving
+
+Considering the `Label<T>` type above again, and a few derives:
+
+```rust
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Label<T> {
+    marker: PhantomData<T>,
+}
+```
+
+we will now get get following `impl`s:
+
+```rust
+impl<T: Copy>       Copy       for Label<T> { /* .. */ }
+impl<T: Clone>      Clone      for Label<T> { /* .. */ }
+impl<T: Debug>      Debug      for Label<T> { /* .. */ }
+impl<T: Default>    Default    for Label<T> { /* .. */ }
+impl<T: PartialEq>  PartialEq  for Label<T> { /* .. */ }
+impl<T: Eq>         Eq         for Label<T> { /* .. */ }
+impl<T: PartialOrd> PartialOrd for Label<T> { /* .. */ }
+impl<T: Ord>        Ord        for Label<T> { /* .. */ }
+impl<T: Hash>       Hash       for Label<T> { /* .. */ }
+```
+
+Notice the bounds on the type parameter `T` in all of these `impl`s. They are
+completely unnecessary and restrictive. Instead, we would like to generate:
+
+```rust
+impl<T> Copy       for Label<T> { /* .. */ }
+impl<T> Clone      for Label<T> { /* .. */ }
+impl<T> Debug      for Label<T> { /* .. */ }
+impl<T> Default    for Label<T> { /* .. */ }
+impl<T> PartialEq  for Label<T> { /* .. */ }
+impl<T> Eq         for Label<T> { /* .. */ }
+impl<T> PartialOrd for Label<T> { /* .. */ }
+impl<T> Ord        for Label<T> { /* .. */ }
+impl<T> Hash       for Label<T> { /* .. */ }
+```
+
+But deriving macros can't generate such `impl`s since it can't be completely
+sure that `PhantomData<T>` really is `::core::marker::PhantomData<T>`.
+Macros can chance it, but that is not reliable, which is why deriving macros
+for derivable standard library traits don't assume getting `PhantomData<T>`
+as a field type in the input token stream really is that type. Instead we
+are faced with having to implement the traits manually for the type, which
+is a waste. If however, a macros gets the following input:
+
+```rust
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Label<T>;
```

Comment:
Here it seems `phantom T` is indeed optional.

Reply:
Yep, that's right.